File: src/main/java/com/mpe85/grampa/rule/RuleContext.java
package com.mpe85.grampa.rule;

import com.mpe85.grampa.input.InputBuffer;

/**
 * Defines a context for parser rules.
 * 
 * @author mpe85
 *
 * @param <T>
 *            the type of the stack elements
 */
public interface RuleContext<T> extends ActionContext<T> {
	
	/**
	 * Gets the underlying input buffer.
	 * 
	 * @return an input buffer
	 */
	InputBuffer getInputBuffer();
	
	/**
	 * Sets the current index inside the parser input.
	 * 
	 * @param currentIndex
	 *            a valid index
	 */
	void setCurrentIndex(int currentIndex);
	
	/**
	 * Advances the current index inside the parser input.
	 * 
	 * @param delta
	 *            number of characters to advance
	 * @return true if the advancing was possible, otherwise false
	 */
	boolean advanceIndex(int delta);
	
	/**
	 * Runs the parser against the input.
	 * 
	 * @return true if the parser successfully matched the input, otherwise false
	 */
	boolean run();
	
	@Override
	RuleContext<T> getParent();
	
	/**
	 * Creates a child context out of the context that is used to run child rules.
	 * 
	 * @param rule
	 *            a child rule for which the child context should be created for
	 * @return a rule contexts
	 */
	RuleContext<T> createChildContext(Rule<T> rule);
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/RuleContext.kt
package com.mpe85.grampa.rule

import com.mpe85.grampa.input.InputBuffer

/**
 * A context for parser rules.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @property[inputBuffer] The underlying input buffer
 * @property[currentIndex] The current index inside the parser input
 * @property[parent] The parent context of the context
 */
interface RuleContext<T> : ActionContext<T> {

  val inputBuffer: InputBuffer
  override var currentIndex: Int
  override val parent: RuleContext<T>?

  /**
   * Advance the current index inside the parser input.
   *
   * @param[delta] The number of characters to advance
   * @return true if the advancing was possible
   */
  fun advanceIndex(delta: Int): Boolean

  /**
   * Run the parser against the input.
   *
   * @return true if the parser successfully matched the input
   */
  fun run(): Boolean

  /**
   * Create a child context out of the context that is used to run child rules.
   *
   * @param[rule] A child rule for which the child context should be created for
   * @return A new rule context
   */
  fun createChildContext(rule: Rule<T>): RuleContext<T>

}
